File: gameserver/src/main/java/l2s/gameserver/model/items/listeners/ItemAugmentationListener.java
package l2s.gameserver.model.items.listeners;

import l2s.gameserver.data.xml.holder.OptionDataHolder;
import l2s.gameserver.model.Playable;
import l2s.gameserver.model.Player;
import l2s.gameserver.model.items.ItemInstance;
import l2s.gameserver.templates.OptionDataTemplate;
import l2s.gameserver.templates.item.ItemTemplate;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Bonux
**/
public final class ItemAugmentationListener extends AbstractOptionDataListener
{
	private static final ItemAugmentationListener _instance = new ItemAugmentationListener();

	public static ItemAugmentationListener getInstance()
	{
		return _instance;
	}

	@Override
	public int onEquip(int slot, ItemInstance item, Playable actor)
	{
		if(!item.isEquipable())
			return 0;

		if(!item.isAugmented())
			return 0;

		if(!actor.isPlayer())
			return 0;

		Player player = actor.getPlayer();

		int flags = 0;

		List<OptionDataTemplate> addedOptionDatas = new ArrayList<OptionDataTemplate>();

		// При несоотвествии грейда аугмент не применяется
		if(item.getTemplate().getType2() != ItemTemplate.TYPE2_WEAPON || player.getExpertiseWeaponPenalty() == 0)
		{
			int[] stats = { item.getVariation1Id(), item.getVariation2Id() };
			for(int i : stats)
			{
				OptionDataTemplate template = OptionDataHolder.getInstance().getTemplate(i);
				if(template == null)
					continue;

				addedOptionDatas.add(template);
			}
		}

		flags |= refreshOptionDatas(actor, item, addedOptionDatas);
		return flags;
	}

	@Override
	public int onRefreshEquip(ItemInstance item, Playable actor)
	{
		return onEquip(item.getEquipSlot(), item, actor);
	}
}

File: gameserver/src/main/java/l2s/gameserver/model/items/listeners/ItemAugmentationListener.kt
package l2s.gameserver.model.items.listeners

import l2s.gameserver.data.xml.holder.OptionDataHolder
import l2s.gameserver.listener.inventory.OnEquipListener
import l2s.gameserver.model.Playable
import l2s.gameserver.model.items.ItemInstance

object ItemAugmentationListener : OnEquipListener {

    override fun onEquip(slot: Int, item: ItemInstance, actor: Playable) {
        if (!item.isEquipable)
            return
        if (!item.isAugmented)
            return

        val player = actor.player ?: return

        // При несоотвествии грейда аугмент не применяется
        if (player.getExpertisePenalty(item) > 0)
            return

        val stats = item.augmentations

        var updateStats = false
        var sendSkillList = false
        for (i in stats) {
            val template = OptionDataHolder.getInstance().getTemplate(i) ?: continue

            if (player.addOptionData(template) != template) {
                updateStats = true
                if (!template.skills.isEmpty())
                    sendSkillList = true
            }
        }

        if (updateStats) {
            if (sendSkillList)
                player.sendSkillList()
            player.sendChanges()
        }
    }

    override fun onUnequip(slot: Int, item: ItemInstance, actor: Playable) {
        if (!item.isEquipable)
            return
        if (!item.isAugmented)
            return

        val player = actor.player ?: return

        val stats = item.augmentations

        var updateStats = false
        var sendSkillList = false
        for (i in stats) {
            val template = player.removeOptionData(i)
            if (template != null) {
                updateStats = true
                if (template.skills.isNotEmpty())
                    sendSkillList = true
            }
        }

        if (updateStats) {
            if (sendSkillList)
                player.sendSkillList()
            player.updateStats()
        }
    }

}